source/Halibut/IDataStreamReceiver.cs

```diff
@@ -1,11 +1,12 @@
 using System;
 using System.IO;
+using System.Threading.Tasks;
 
 namespace Halibut
 {
     public interface IDataStreamReceiver
     {
-        void SaveTo(string filePath);
-        void Read(Action<Stream> reader);
+        Task SaveTo(string filePath);
+        Task Read(Func<Stream, Task> reader);
```

Comment:
is the assumption here that nothing else needs to be accessed in the lambda but the stream? If not I'd say this probably requires an overload which accepts a state object if you want to avoid delegate allocations due to closure capturing.